src/trusted-device/Security/TwoFactor/Trusted/JwtTokenEncoder.php

```diff
@@ -16,51 +19,56 @@ class JwtTokenEncoder
     public const CLAIM_VERSION = 'vsn';
 
     /**
-     * @var Sha256
+     * @var Configuration
      */
-    private $signer;
-
-    /**
-     * @var string
-     */
-    private $applicationSecret;
+    private $configuration;
 
     public function __construct(string $applicationSecret)
     {
-        $this->signer = new Sha256();
-        $this->applicationSecret = $applicationSecret;
+        $this->configuration = Configuration::forSymmetricSigner(new Sha256(), InMemory::plainText($applicationSecret));
+        $this->configuration->setValidationConstraints(new SignedWith($this->configuration->signer(), $this->configuration->signingKey()));
     }
 
     public function generateToken(string $username, string $firewallName, int $version, \DateTimeInterface $validUntil): Token
     {
-        $builder = (new Builder())
-            ->setIssuedAt(time())
-            ->setExpiration($validUntil->getTimestamp())
-            ->set(self::CLAIM_USERNAME, $username)
-            ->set(self::CLAIM_FIREWALL, $firewallName)
-            ->set(self::CLAIM_VERSION, $version)
-            ->sign($this->signer, $this->applicationSecret);
+        $validUntil = $validUntil;
+        if ($validUntil instanceof \DateTime) {
+            $validUntil = \DateTimeImmutable::createFromMutable($validUntil);
+        } elseif (!$validUntil instanceof \DateTimeImmutable) {
+            throw new \InvalidArgumentException(sprintf('The 4th argument to "%s" must be instance of either a "DateTime" or "DateTimeImmutable", "%s" given.', __METHOD__, \get_class($validUntil)));
+        }
 
-        return $builder->getToken();
+        $builder = $this->configuration->builder()
+            ->issuedAt(new \DateTimeImmutable())
+            ->expiresAt($validUntil)
+            ->withClaim(self::CLAIM_USERNAME, $username)
+            ->withClaim(self::CLAIM_FIREWALL, $firewallName)
+            ->withClaim(self::CLAIM_VERSION, $version);
+
+        return $builder->getToken($this->configuration->signer(), $this->configuration->signingKey());
     }
 
     public function decodeToken(string $token): ?Token
     {
         try {
-            $token = (new Parser())->parse($token);
-        } catch (\InvalidArgumentException $e) {
+            $token = $this->configuration->parser()->parse($token);
+        } catch (Exception $e) {
             return null; // Could not decode token
         }
 
+        if (!$token instanceof Plain) {
+            return null;
+        }
```

Comment:
I've added the check for `Plain` in this class. Although it's not strictly necessary, all code using `decodeToken()` requires it to be a plain token. Do you also want me to change the return type to `?Plain`? (this would remove the exception in `src/trusted-device/Security/TwoFactor/Trusted/TrustedDeviceTokenEncoder.php:32`

Reply:
As I've said in the other comment, I consider these classes implementation details rather than an extension points. So I'm fine with changing all signatures from `Token` to the `Plain` class.

Reply:
Alright, I would recommend you to flag them with `@internal` or `@final` then. That communicates your considerations with these classes to everyone using your libraries (I've always been amazed by which classes of OSS projects people extend/override).

On top of this, the Symfony DebugClassLoader will also warn when debug=true people when they extend from an `@final` class or use an `@internal` class.

Reply:
Yea, that same thought also came to my mind when I read your comment. You know know what crazy stuff people do with software. I just wasn't sure how to do it properly, so thanks for explaining.